includes/data-stores/class-wc-product-variation-data-store-cpt.php

```diff
@@ -0,0 +1,254 @@
+			'length'            => get_post_meta( $id, '_length', true ),
+			'width'             => get_post_meta( $id, '_width', true ),
+			'height'            => get_post_meta( $id, '_height', true ),
+			'tax_class'         => get_post_meta( $id, '_tax_class', true ),
+		) );
+
+		if ( $product->is_on_sale() ) {
+			$product->set_price( $product->get_sale_price() );
+		} else {
+			$product->set_price( $product->get_regular_price() );
+		}
+
+		$product->set_parent_data( array(
+			'sku'            => get_post_meta( $product->get_parent_id(), '_sku', true ),
+			'manage_stock'   => get_post_meta( $product->get_parent_id(), '_manage_stock', true ),
+			'backorders'     => get_post_meta( $product->get_parent_id(), '_backorders', true ),
+			'stock_quantity' => get_post_meta( $product->get_parent_id(), '_stock', true ),
+			'weight'         => get_post_meta( $product->get_parent_id(), '_weight', true ),
+			'length'         => get_post_meta( $product->get_parent_id(), '_length', true ),
+			'width'          => get_post_meta( $product->get_parent_id(), '_width', true ),
+			'height'         => get_post_meta( $product->get_parent_id(), '_height', true ),
+			'tax_class'      => get_post_meta( $product->get_parent_id(), '_tax_class', true ),
+			'image_id'       => get_post_thumbnail_id( $product->get_parent_id() ),
+		) );
+
+		return $product;
+	}
+
+	/**
+	 * For all stored terms in all taxonomies, save them to the DB.
+	 *
+	 * @since 2.7.0
+	 * @param WC_Product
+	 */
+	protected function update_terms( $product ) {
+		wp_set_post_terms( $product->get_id(), array( $product->get_shipping_class_id( 'edit' ) ), 'product_shipping_class', false );
+	}
+
+	/**
+	 * Update attribute meta values.
+	 *
+	 * @since 2.7.0
+	 * @param WC_Product
+	 */
+	protected function update_attributes( $product ) {
+		global $wpdb;
+		$attributes             = $product->get_attributes();
+		$updated_attribute_keys = array();
+		foreach ( $attributes as $key => $value ) {
+			update_post_meta( $product->get_id(), 'attribute_' . $key, $value );
+			$updated_attribute_keys[] = 'attribute_' . $key;
+		}
+
+		// Remove old taxonomies attributes so data is kept up to date - first get attribute key names.
+		$delete_attribute_keys = $wpdb->get_col( $wpdb->prepare( "SELECT meta_key FROM {$wpdb->postmeta} WHERE meta_key LIKE 'attribute_%%' AND meta_key NOT IN ( '" . implode( "','", array_map( 'esc_sql', $updated_attribute_keys ) ) . "' ) AND post_id = %d;", $product->get_id() ) );
+
+		foreach ( $delete_attribute_keys as $key ) {
+			delete_post_meta( $product->get_id(), $key );
+		}
+	}
+
+	/**
+	 * Helper method that updates all the post meta for a product based on it's settings in the WC_Product class.
+	 *
+	 * @since 2.7.0
+	 * @param WC_Product
+	 * @return WC_Product
+	 */
+	public function update_post_meta( $product ) {
+		update_post_meta( $product->get_id(), '_variation_description', $product->get_description() );
+		$product = parent::update_post_meta( $product );
```

Comment:
Would it be cleaner to pass by reference rather than pass the $product object around? I don't know the answer - I just notice you used a mixture of both approaches throughout the PR.